internal/app/secretless/handlers/mysql/handler.go

```diff
@@ -11,48 +11,127 @@ import (
+	sequenceId byte
+}
+
+
+func NewMySQLClientConnection(conn net.Conn) *MySQLConnection {
+	return &MySQLConnection{sequenceId: 0, conn: conn}
+}
+
+func NewMySQLBackendConnection(conn net.Conn) *MySQLConnection {
+	return &MySQLConnection{sequenceId: 1, conn: conn}
+}
+
+func (c *MySQLConnection) RawConnection() net.Conn {
+	return c.conn
+}
+
+func (c *MySQLConnection) SetConnection(conn net.Conn) {
+	c.conn = conn
+}
+
+func (c *MySQLConnection) write(pkt mysqlPacket) error {
+	pkt.SetSequenceId(c.sequenceId)
+	c.sequenceId++
+	if _, err := protocol.WritePacket(pkt, c.conn); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *MySQLConnection) read() (mysqlPacket, error) {
+	pkt, err := protocol.ReadPacket(c.conn)
+	if err != nil {
+		return nil, err
+	}
+
+	mysqlPkt := mysqlPacket(pkt)
+	c.sequenceId = (&mysqlPkt).SequenceId()
+	c.sequenceId++
+	return pkt, nil
+}
+
+func (cd *ConnectionDetails) Address() string {
+	return cd.Host + ":" + strconv.FormatUint(uint64(cd.Port), 10)
 }
 
 // Handler connects a client to a backend. It uses the handler Config and Providers to
-// establish the BackendConfig, which is used to make the Backend connection. Then the data
+// establish the ConnectionDetails, which is used to make the Backend connection. Then the data
 // is transferred bidirectionally between the Client and Backend.
 //
 // Handler requires "host", "port", "username" and "password" credentials.
+//
 type Handler struct {
-	// SequenceIds keep track of packet no. for client and backend
-	// MySQL uses this to ensure packets arrive in order
-	// values indicate appropriate value to use on subsequent write
-	clientSequenceId byte
-	backendSequenceId byte
+
+	// The connections are decorated versions of net.Conn that allow us
+	// to do read/writes according to the MySQL protocol.  Protocol level
+	// details are thus encapsulated.  Within the MySQL code, we _only_
+	// deal with these decorated versions.
+
+	mySQLClientConn *MySQLConnection
+	mySQLBackendConn *MySQLConnection
 	plugin_v1.BaseHandler
-	BackendConfig *BackendConfig
-}
-
-func (h *Handler) abort(err error) {
-	if h.GetClientConnection() != nil {
-		var mysqlError *protocol.Error
-		var ok bool
-		if mysqlError, ok = err.(*protocol.Error); !ok {
-			mysqlError = &protocol.Error{
-				Code:     protocol.CRUnknownError,
-				SQLSTATE: protocol.ErrorCodeInternalError,
-				Message:  err.Error(),
-			}
-		}
+	connectionDetails *ConnectionDetails
+}
+
+// If the error is not already a MySQL protocol error, then wrap it in an
+// "Unknown" MySQL protocol error, because the client understands only those.
+//
+func (h *Handler) sendErrorToClient(err error) {
+	mysqlError, isProtocolErr := err.(*protocol.Error)
+
+	if !isProtocolErr {
+		mysqlError = genericProtocolError(err)
+	}
+
+	// TODO: log something if this fails
+	//
+	h.mySQLClientConn.write(mysqlError.GetMessage())
+}
 
-		h.ClientWrite(mysqlError.GetMessage())
+func genericProtocolError(goErr error) *protocol.Error {
```

Comment:
this feels like it belongs somewhere in the protocol (protocol/error.go? protocol/protocol.go?) and not here